### ISSUE TITLE
Crash on rotate

### ISSUE THREAD

**Issue by avuton** [stefan-niedermann/nextcloud-notes#1239]
Hello,

This bug has been hitting me for the last 2 or 3 weeks, I think. Happens when I rotate 90 then I rotate the screen back to the original position. I did not find it in the closed issues, sorry if I missed it (I have a tendancy to :()

**Describe the bug**
Crash

**To Reproduce**
1. Open any note
2. Go from portrait to landscape to portrait with autorotate.

**Expected behavior**
Not crash

**Screenshots**

**Smartphone (please complete the following information):**
 - Nextcloud Notes-Version (android app): 3.4.3
 - F-Droid or Play Store: F-Droid 
 - Android-Version: 4.4.268-lineage+(b5a96127f4)
 - Device: OnePlus5T

**Server**
 - Nextcloud version: 21.0.2
 - Nextcloud Notes version (server app): 4.0.4

**Stacktrace**

```
App Version: 3.4.3
App Version Code: 3004003
App Flavor: fdroid

Files App Version Code: 30160090

---

OS Version: 4.4.268-lineage+(b5a96127f4)
OS API Level: 30
Device: OnePlus5T
Manufacturer: OnePlus
Model (and Product): ONEPLUS A5010 (OnePlus5T)

---

java.lang.RuntimeException: Unable to start activity ComponentInfo{it.niedermann.owncloud.notes/it.niedermann.owncloud.notes.edit.EditNoteActivity}: java.lang.RuntimeException: Window couldn't find content container view
	at android.app.ActivityThread.performLaunchActivity(ActivityThread.java:3431)
	at android.app.ActivityThread.handleLaunchActivity(ActivityThread.java:3595)
	at android.app.ActivityThread.handleRelaunchActivityInner(ActivityThread.java:5451)
	at android.app.ActivityThread.handleRelaunchActivity(ActivityThread.java:5357)
	at android.app.servertransaction.ActivityRelaunchItem.execute(ActivityRelaunchItem.java:69)
	at android.app.servertransaction.TransactionExecutor.executeCallbacks(TransactionExecutor.java:135)
	at android.app.servertransaction.TransactionExecutor.execute(TransactionExecutor.java:95)
	at android.app.ActivityThread$H.handleMessage(ActivityThread.java:2066)
	at android.os.Handler.dispatchMessage(Handler.java:106)
	at android.os.Looper.loop(Looper.java:223)
	at android.app.ActivityThread.main(ActivityThread.java:7660)
	at java.lang.reflect.Method.invoke(Native Method)
	at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:592)
	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:947)
Caused by: java.lang.RuntimeException: Window couldn't find content container view
	at com.android.internal.policy.PhoneWindow.generateLayout(PhoneWindow.java:2636)
	at com.android.internal.policy.PhoneWindow.installDecor(PhoneWindow.java:2695)
	at com.android.internal.policy.PhoneWindow.getDecorView(PhoneWindow.java:2118)
	at androidx.appcompat.app.AppCompatDelegateImpl.createSubDecor(AppCompatDelegateImpl.java:864)
	at androidx.appcompat.app.AppCompatDelegateImpl.ensureSubDecor(AppCompatDelegateImpl.java:806)
	at androidx.appcompat.app.AppCompatDelegateImpl.initWindowDecorActionBar(AppCompatDelegateImpl.java:547)
	at androidx.appcompat.app.AppCompatDelegateImpl.getSupportActionBar(AppCompatDelegateImpl.java:534)
	at androidx.appcompat.app.AppCompatDelegateImpl.invalidateOptionsMenu(AppCompatDelegateImpl.java:1217)
	at androidx.appcompat.app.AppCompatActivity.supportInvalidateOptionsMenu(AppCompatActivity.java:273)
	at androidx.fragment.app.FragmentActivity$HostCallbacks.onSupportInvalidateOptionsMenu(FragmentActivity.java:774)
	at androidx.fragment.app.Fragment.setHasOptionsMenu(Fragment.java:1275)
	at it.niedermann.owncloud.notes.edit.BaseNoteFragment.onCreate(BaseNoteFragment.java:146)
	at it.niedermann.owncloud.notes.edit.NoteEditFragment.onCreate(NoteEditFragment.java:65)
	at androidx.fragment.app.Fragment.performCreate(Fragment.java:2949)
	at androidx.fragment.app.FragmentStateManager.create(FragmentStateManager.java:475)
	at androidx.fragment.app.FragmentStateManager.moveToExpectedState(FragmentStateManager.java:278)
	at androidx.fragment.app.FragmentStore.moveToExpectedState(FragmentStore.java:112)
	at androidx.fragment.app.FragmentManager.moveToState(FragmentManager.java:1647)
	at androidx.fragment.app.FragmentManager.dispatchStateChange(FragmentManager.java:3126)
	at androidx.fragment.app.FragmentManager.dispatchCreate(FragmentManager.java:3059)
	at androidx.fragment.app.FragmentController.dispatchCreate(FragmentController.java:240)
	at androidx.fragment.app.FragmentActivity.onCreate(FragmentActivity.java:276)
	at androidx.appcompat.app.AppCompatActivity.onCreate(AppCompatActivity.java:115)
	at it.niedermann.owncloud.notes.LockedActivity.onCreate(LockedActivity.java:25)
	at it.niedermann.owncloud.notes.edit.EditNoteActivity.onCreate(EditNoteActivity.java:63)
	at android.app.Activity.performCreate(Activity.java:8000)
	at android.app.Activity.performCreate(Activity.java:7984)
	at android.app.Instrumentation.callActivityOnCreate(Instrumentation.java:1309)
	at android.app.ActivityThread.performLaunchActivity(ActivityThread.java:3404)
	... 13 more

```

**Comment by stefan-niedermann** [comment]
Okay, i google around a bit and i doubt that this is caused by the Notes app itself. I believe it is an issue of the AndroidX library which we updated recently.
Which Android version does your currently installed LineageOS match?

I will try to set up a build with the previous AndroidX version. Are you willing to test this build?

Since i can't reproduce the problem, i rely on your help to fix it 🙂

**Comment by avuton** [comment]
Super happy to test. Lineageos lineage_dumpling-userdebug 11 RQ2A.210505.003 b5a96127f4 (May 25, 2021).

**Comment by stefan-niedermann** [comment]
Here we go!

Do you notice any difference with [this build](https://github.com/stefan-niedermann/nextcloud-notes/suites/2862752911/artifacts/63949467)?
You can safely install and test the `.apk` parallel to the released version.

**Comment by stefan-niedermann** [comment]
ping @aslmx & @avuton can somebody test this build please?

**Comment by avuton** [comment]
> ping @aslmx & @avuton can somebody test this build please?

Sorry! I still have a crash. I dont know if it's related and haven't had time to collect a proper crash report. I should have time within 24h.

**Comment by stefan-niedermann** [comment]
Since you noted that this did not happen in the past, it will be a valuable information which was the first version that had this crash. You can install older versions via F-Droid to try it out - since i am not able to repoduce the issue, i will need the exact version which introduce the problem to fix it...

**Comment by avuton** [comment]
> Since you noted that this did not happen in the past, it will be a valuable information which was the first version that had this crash. You can install older versions via F-Droid to try it out - since i am not able to repoduce the issue, i will need the exact version which introduce the problem to fix it...

You got it, or I'll bisect.

**Comment by avuton** [comment]
I am working on this, unfortunately, there are a lot of commits I have to skip for various reasons, and I have time commitments keeping me from getting to the breaking commit. Hope to have something tomorrow.

**Comment by aslmx** [comment]
I just took the time to uninstall and reinstall 3.4.2 from f-droid (that is the oldest i can get today).

my steps to reproduce were:
1. Open an existing note
2. Change something
3. rotate device to landscape
4. rotate device to portrait
5. -> Crashes

(sometimes takes a few iterations of step # 2 to # 4)

It took me a few edits/rotates but even in 3.4.2 it happens already And i think 3.4.3 has it, too.

I can't promise, but maybe over the weekend I'd have some time - maybe. So if you have older binaries i can try to reproduce.

**Comment by avuton** [comment]
f31a1756e4bfb3cb2b6bbcf5f6cf5fd1f2133699 is the first bad commit
commit f31a1756e4bfb3cb2b6bbcf5f6cf5fd1f2133699
Author: Stefan Niedermann <info@niedermann.it>
Date:   Fri Apr 2 19:58:50 2021 +0200

    Merge scrollY logic from master

 .../owncloud/notes/edit/BaseNoteFragment.java      | 47 +++++++++++-----------
 .../owncloud/notes/edit/NoteEditFragment.java      |  7 ++++
 .../owncloud/notes/edit/NotePreviewFragment.java   | 14 ++++++-
 3 files changed, 43 insertions(+), 25 deletions(-)

**Comment by stefan-niedermann** [comment]
That's very precise, thank you! Will have a look.

**Comment by larena1** [comment]
Have you had any success in fixing this? Notes crashing very often unfortunately.

**Comment by stefan-niedermann** [comment]
I currently suspect the problem to happen if the wanted `scrollY` is off screen in edit or preview mode in case that the last stored state from the previous mode was longer (for example the scrollable area has a different height in preview with large images or edit mode with code blocks and long links).

@larena1 this is not the first time ([1](https://github.com/stefan-niedermann/nextcloud-notes/issues/549#issuecomment-650800264), [2](https://github.com/stefan-niedermann/nextcloud-notes/issues/549#issuecomment-705143447)) that i tell you: Please use **[GitHub reactions](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/)** 👍 to show that you are affected by the same issue. Please don't comment if you have no relevant information to add. This spams all people who subscribed to an issue and makes it unreadable for developers.

I am tired of repeating myself that often. I mean what do you think? I secretly found the solution but didn't tell anyone? 😉 If i have news, i will post it here if course. If not: Feel free to contribute a solution yourself. But please stop commenting if you have nothing to contribute.

**Comment by avuton** [comment]
> I currently suspect the problem to happen if the wanted `scrollY` is off screen in edit or preview mode in case that the last stored state from the previous mode was longer (for example the scrollable area has a different height in preview with large images or edit mode with code blocks and long links).

I'm not sure this completely negates that theory in any way, but I'll add some details:

* I have tried to isolate it to a smaller piece of code, which has failed.
* My note for testing this bug is, literally, the letter 'S' in a note, but I can duplicate it with other notes.
* Sometimes, it doesn't seem to "completely" crash, the screen blanks for a second (a hang?) and I get kicked out of the note and put to the main menu (this might also be a unreported crash, I didn't check the logcat). If I go back to the note and rotate a few more times, it'll usually crash.

**Comment by stefan-niedermann** [comment]
Yep, i think i tracked it down.

[Changing from (the deprecated) `onActivityCreated` to `onCreate`](https://github.com/stefan-niedermann/nextcloud-notes/commit/f31a1756e4bfb3cb2b6bbcf5f6cf5fd1f2133699#diff-db760cf7a629cf2c0756f9ce530d2ff650d06e531299877830fd0f283563e3bfR97) might cause the issue and explain why it does not happen everytime but only sometimes.

The note gets loaded asynchronously and *sometimes* this takes so long, that the view has been created, before the content will be set. *Sometimes* though, it will be so fast, that the view has not been created yet.

After [changing the implementation from `onCreate` to `onViewCreated`](https://github.com/stefan-niedermann/nextcloud-notes/commit/a747c5d045a7b6c64ae36febebc9911da6b2158c) i haven't been able to reproduce the problem anymore.

It would be great if @avuton or @larena1 can verify that this specific issue has been fixed by installing [this debug build](https://github.com/stefan-niedermann/nextcloud-notes/suites/2970373474/artifacts/66967953) (it can be installed in parallel to the published version).

**Comment by stefan-niedermann** [comment]
Pretty ironic that old and slow devices will probably have less problems than fast modern devices due to this bug :smile:

**Comment by avuton** [comment]
I'm reasonably confident that build has fixed it, thank you so much for your time!

**Comment by stefan-niedermann** [comment]
Thank you much for the extremily helpful bad commit you identified! And for the detailled bug report 🙂 

I will close this issue, the fix will be in `v3.4.9`. Feel free to reopen if the issue occurs again with this or a higher version :)

**Comment by IngoBelka** [comment]
Hello,
on my Huawei MediaPad 5 with Android 9 the rotation bug is still present.

```
App Version: 3.4.8
App Version Code: 3004008
App Flavor: fdroid

Files App Version Code: 30160190

---

OS Version: 4.9.148(9.1.0.335C432)
OS API Level: 28
Device: HWSHT
Manufacturer: HUAWEI
Model (and Product): SHT-W09 (SHT-W09)

---

java.lang.RuntimeException: Unable to start activity ComponentInfo{it.niedermann.owncloud.notes/it.niedermann.owncloud.notes.edit.EditNoteActivity}: java.lang.NullPointerException: Attempt to invoke virtual method 'void android.view.ViewGroup.removeAllViews()' on a null object reference
	at android.app.ActivityThread.performLaunchActivity(ActivityThread.java:3430)
	at android.app.ActivityThread.handleLaunchActivity(ActivityThread.java:3614)
	at android.app.ActivityThread.handleRelaunchActivityInner(ActivityThread.java:5527)
	at android.app.ActivityThread.handleRelaunchActivity(ActivityThread.java:5418)
	at android.app.servertransaction.ActivityRelaunchItem.execute(ActivityRelaunchItem.java:69)
	at android.app.servertransaction.TransactionExecutor.executeCallbacks(TransactionExecutor.java:108)
	at android.app.servertransaction.TransactionExecutor.execute(TransactionExecutor.java:68)
	at android.app.ActivityThread$H.handleMessage(ActivityThread.java:2199)
	at android.os.Handler.dispatchMessage(Handler.java:112)
	at android.os.Looper.loop(Looper.java:216)
	at android.app.ActivityThread.main(ActivityThread.java:7625)
	at java.lang.reflect.Method.invoke(Native Method)
	at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:524)
	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:987)
Caused by: java.lang.NullPointerException: Attempt to invoke virtual method 'void android.view.ViewGroup.removeAllViews()' on a null object reference
	at androidx.appcompat.app.AppCompatDelegateImpl.setContentView(AppCompatDelegateImpl.java:689)
	at androidx.appcompat.app.AppCompatActivity.setContentView(AppCompatActivity.java:201)
	at it.niedermann.owncloud.notes.edit.EditNoteActivity.onCreate(EditNoteActivity.java:77)
	at android.app.Activity.performCreate(Activity.java:7458)
	at android.app.Activity.performCreate(Activity.java:7448)
	at android.app.Instrumentation.callActivityOnCreate(Instrumentation.java:1286)
	at android.app.ActivityThread.performLaunchActivity(ActivityThread.java:3409)
	... 13 more

```

**Comment by stefan-niedermann** [comment]
I think we might have two independent issues here (See [two different stacktraces](https://github.com/stefan-niedermann/nextcloud-notes/issues/1239#issuecomment-852744193)).

Probably they are caused by a similar mistake, so i'll have another look into the `EditNoteActivity` and the edit and preview fragments.
@IngoBelka can you tell us whether this is only happening in edit, preview or both modes?

**Comment by stefan-niedermann** [comment]
Ah, i just noticed that you are using `v3.4.8`, the bugfix will be shipped with `v3.4.9`, so just be a bit more patient. F-Droid alwys takes some time to build and publish new versions 🙂 .
